File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/live/fragment/dialog/EarBackFragment.java
package io.agora.scene.ktv.live.fragment.dialog;

import android.annotation.SuppressLint;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RadioGroup;
import android.widget.SeekBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.agora.scene.base.component.BaseViewBindingFragment;
import io.agora.scene.base.component.OnButtonClickListener;
import io.agora.scene.ktv.R;
import io.agora.scene.ktv.databinding.KtvDialogEarbackSettingBinding;
import io.agora.scene.ktv.widget.MusicSettingBean;
import io.agora.scene.widget.dialog.CommonDialog;

public class EarBackFragment extends BaseViewBindingFragment<KtvDialogEarbackSettingBinding> {
    public static final String TAG = "EarBackFragment";

    private final MusicSettingBean mSetting;
    private CommonDialog switchEarModeDialog;

    public EarBackFragment(MusicSettingBean mSetting) {
        this.mSetting = mSetting;
        mSetting.setEarPhoneCallback(new MusicSettingBean.EarPhoneCallback() {
            @Override
            public void onHasEarPhoneChanged(boolean hasEarPhone) {
                updateEarPhoneStatus(hasEarPhone);
            }

            @Override
            public void onEarMonitorDelay(int earsBackDelay) {
                updateEarPhoneDelay(earsBackDelay);
            }
        });
    }

    @NonNull
    @Override
    protected KtvDialogEarbackSettingBinding getViewBinding(@NonNull LayoutInflater layoutInflater, @Nullable ViewGroup viewGroup) {
        return KtvDialogEarbackSettingBinding.inflate(layoutInflater);
    }

    @Override
    public void initListener() {
        super.initListener();

        updateEarPhoneStatus(mSetting.hasEarPhone());
        if (this.mSetting.isEar()) {
            getBinding().vSettingMark.setVisibility(View.INVISIBLE);
            getBinding().vPingMark.setVisibility(View.INVISIBLE);
        } else {
            getBinding().vSettingMark.setVisibility(View.VISIBLE);
            getBinding().vPingMark.setVisibility(View.VISIBLE);
            getBinding().vSettingMark.setOnClickListener(v -> {});
        }

        getBinding().btEarBackDown.setOnClickListener(v -> tuningEarVolume(false));
        getBinding().btEarBackUp.setOnClickListener(v -> tuningEarVolume(true));
        getBinding().sbEarBack.setProgress(this.mSetting.getEarBackVolume());
        getBinding().sbEarBack.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                mSetting.setEarBackVolume(progress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        if (this.mSetting.getEarBackMode() == 0) {
            getBinding().rgMode.check(R.id.tvModeAuto);
        } else if (this.mSetting.getEarBackMode() == 1) {
            getBinding().rgMode.check(R.id.tvModeOpenSL);
        } else {
            getBinding().rgMode.check(R.id.tvModeOboe);
        }
        getBinding().rgMode.setOnCheckedChangeListener((group, checkedId) -> {
            if (checkedId == R.id.tvModeAuto) {
                mSetting.setEarBackMode(0);
            } else if (checkedId == R.id.tvModeOpenSL) {
                showSwitchEarModeDialog(1);
            } else {
                showSwitchEarModeDialog(2);
            }
        });

        updateEarPhoneDelay(mSetting.getEarBackDelay());
    }

    private void updateEarPhoneStatus(boolean hasEarPhone) {
        if (hasEarPhone) {
            // 检测到耳机
            getBinding().cbSwitch.setChecked(mSetting.isEar());
            getBinding().cbSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
                mSetting.setEar(isChecked);
                if (isChecked) {
                    getBinding().vSettingMark.setVisibility(View.INVISIBLE);
                    getBinding().vPingMark.setVisibility(View.INVISIBLE);
                } else {
                    getBinding().vSettingMark.setVisibility(View.VISIBLE);
                    getBinding().vPingMark.setVisibility(View.VISIBLE);
                    getBinding().vSettingMark.setOnClickListener(v -> {});
                }
            });
            getBinding().tvTips.setVisibility(View.VISIBLE);
            getBinding().tvTipsNoEarPhone.setVisibility(View.INVISIBLE);
        } else {
            // 未检测到耳机
            getBinding().cbSwitch.setEnabled(false);
            mSetting.setEar(false);
            getBinding().cbSwitch.setChecked(false);
            getBinding().tvTips.setVisibility(View.INVISIBLE);
            getBinding().tvTipsNoEarPhone.setVisibility(View.VISIBLE);
        }
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    private void updateEarPhoneDelay(int earPhoneDelay) {
        getBinding().pbPing.setProgress(earPhoneDelay);
        getBinding().tvPing.setText(earPhoneDelay + "ms");
        if (earPhoneDelay <= 50) {
            getBinding().pbPing.setProgressDrawable(getBinding().getRoot().getResources().getDrawable(R.drawable.ktv_ear_ping_progress_good));
        } else if (earPhoneDelay <= 100) {
            getBinding().pbPing.setProgressDrawable(getBinding().getRoot().getResources().getDrawable(R.drawable.ktv_ear_ping_progress));
        } else {
            getBinding().pbPing.setProgressDrawable(getBinding().getRoot().getResources().getDrawable(R.drawable.ktv_ear_ping_progress_bad));
        }
    }

    private void tuningEarVolume(boolean volumeUp) {
        int earBackVolume = this.mSetting.getEarBackVolume();
        if (volumeUp) {
            earBackVolume += 1;
        } else {
            earBackVolume -= 1;
        }
        if (earBackVolume > 100)
            earBackVolume = 100;

        if (earBackVolume < 0)
            earBackVolume = 0;
        if (earBackVolume != this.mSetting.getEarBackVolume()) {
            this.mSetting.setEarBackVolume(earBackVolume);
        }
        getBinding().sbEarBack.setProgress(earBackVolume);
    }

    private void showSwitchEarModeDialog(int mode) {
        if (switchEarModeDialog == null) {
            switchEarModeDialog = new CommonDialog(getContext());

            switchEarModeDialog.setDialogTitle("提示");
            if (mode == 1) {
                switchEarModeDialog.setDescText("切换后将强制使用OpenSL模式，\n确认？");
            } else if (mode == 2) {
                switchEarModeDialog.setDescText("切换后将强制使用Oboe模式，\n确认？");
            }
            switchEarModeDialog.setDialogBtnText(getString(R.string.ktv_cancel), getString(R.string.ktv_confirm));
            switchEarModeDialog.setOnButtonClickListener(new OnButtonClickListener() {
                @Override
                public void onLeftButtonClick() {
                    if (mSetting.getEarBackMode() == 0) {
                        getBinding().rgMode.check(R.id.tvModeAuto);
                    } else if (mSetting.getEarBackMode() == 1) {
                        getBinding().rgMode.check(R.id.tvModeOpenSL);
                    } else {
                        getBinding().rgMode.check(R.id.tvModeOboe);
                    }
                }

                @Override
                public void onRightButtonClick() {
                    mSetting.setEarBackMode(mode);
                }
            });
        }
        switchEarModeDialog.show();
    }
}

File: Android/scenes/cantata/src/main/java/io/agora/scene/cantata/ui/fragment/EarBackFragment.kt
package io.agora.scene.cantata.ui.fragment

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.RadioGroup
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import io.agora.scene.base.component.BaseViewBindingFragment
import io.agora.scene.base.component.OnButtonClickListener
import io.agora.scene.cantata.R
import io.agora.scene.cantata.databinding.CantataDialogEarbackSettingBinding
import io.agora.scene.cantata.databinding.CantataFragmentSongListBinding
import io.agora.scene.cantata.ui.activity.RoomLivingActivity
import io.agora.scene.cantata.ui.dialog.EarPhoneCallback
import io.agora.scene.cantata.ui.dialog.MusicSettingBean
import io.agora.scene.widget.dialog.CommonDialog

class EarBackFragment constructor(private val mSetting: MusicSettingBean) :
    BaseViewBindingFragment<CantataDialogEarbackSettingBinding>() {

    companion object {
        const val TAG = "EarBackFragment"
    }

    private var switchEarModeDialog: CommonDialog? = null


    init {
        mSetting.mEarPhoneCallback = object : EarPhoneCallback {
            override fun onHasEarPhoneChanged(hasEarPhone: Boolean) {
                view?.post { updateEarPhoneStatus(hasEarPhone) }
            }

            override fun onEarMonitorDelay(earsBackDelay: Int) {
                view?.post { updateEarPhoneDelay(earsBackDelay) }
            }
        }
    }

    override fun getViewBinding(layoutInflater: LayoutInflater, viewGroup: ViewGroup?): CantataDialogEarbackSettingBinding {
        return CantataDialogEarbackSettingBinding.inflate(layoutInflater)
    }


    override fun initListener() {
        super.initListener()
        updateEarPhoneStatus(mSetting.hasEarPhone)
        if (mSetting.isEar()) {
            binding.vSettingMark.visibility = View.INVISIBLE
            binding.vPingMark.visibility = View.INVISIBLE
        } else {
            binding.vSettingMark.visibility = View.VISIBLE
            binding.vPingMark.visibility = View.VISIBLE
            binding.vSettingMark.setOnClickListener { v: View? -> }
        }
        binding.btEarBackDown.setOnClickListener { v: View? -> tuningEarVolume(false) }
        binding.btEarBackUp.setOnClickListener { v: View? -> tuningEarVolume(true) }
        binding.sbEarBack.progress = mSetting.earBackVolume
        binding.sbEarBack.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                mSetting.earBackVolume = progress
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        when (mSetting.earBackMode) {
            0 -> {
                binding.rgMode.check(R.id.tvModeAuto)
            }
            1 -> {
                binding.rgMode.check(R.id.tvModeOpenSL)
            }
            else -> {
                binding.rgMode.check(R.id.tvModeOboe)
            }
        }
        binding.rgMode.setOnCheckedChangeListener { group: RadioGroup?, checkedId: Int ->
            when (checkedId) {
                R.id.tvModeAuto -> {
                    mSetting.earBackMode = 0
                }
                R.id.tvModeOpenSL -> {
                    showSwitchEarModeDialog(1)
                }
                else -> {
                    showSwitchEarModeDialog(2)
                }
            }
        }
        updateEarPhoneDelay(mSetting.earBackDelay)
        binding.ivBackIcon.setOnClickListener { view: View? -> (requireActivity() as RoomLivingActivity).closeMusicSettingsDialog() }
    }

    private fun updateEarPhoneStatus(hasEarPhone: Boolean) {
        if (hasEarPhone) {
            // 检测到耳机
            binding.cbSwitch.isEnabled = true
            binding.cbSwitch.isChecked = mSetting.isEar()
            binding.cbSwitch.setOnCheckedChangeListener { buttonView: CompoundButton?, isChecked: Boolean ->
                mSetting.setEar(isChecked)
                if (isChecked) {
                    binding.vSettingMark.visibility = View.INVISIBLE
                    binding.vPingMark.visibility = View.INVISIBLE
                } else {
                    binding.vSettingMark.visibility = View.VISIBLE
                    binding.vPingMark.visibility = View.VISIBLE
                    binding.vSettingMark.setOnClickListener { v: View? -> }
                }
            }
            binding.tvTips.visibility = View.VISIBLE
            binding.tvTipsNoEarPhone.visibility = View.INVISIBLE
        } else {
            // 未检测到耳机
            binding.cbSwitch.isEnabled = false
            mSetting.setEar(false)
            binding.cbSwitch.isChecked = false
            binding.tvTips.visibility = View.INVISIBLE
            binding.tvTipsNoEarPhone.visibility = View.VISIBLE
        }
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    private fun updateEarPhoneDelay(earPhoneDelay: Int) {
        binding.pbPing.progress = earPhoneDelay
        binding.tvPing.text = earPhoneDelay.toString() + "ms"
        if (earPhoneDelay <= 50) {
            binding.pbPing.progressDrawable =
                binding.root.resources.getDrawable(R.drawable.cantata_ear_ping_progress_good)
        } else if (earPhoneDelay <= 100) {
            binding.pbPing.progressDrawable =
                binding.root.resources.getDrawable(R.drawable.cantata_ear_ping_progress)
        } else {
            binding.pbPing.progressDrawable =
                binding.root.resources.getDrawable(R.drawable.cantata_ear_ping_progress_bad)
        }
    }

    private fun tuningEarVolume(volumeUp: Boolean) {
        var earBackVolume = mSetting.earBackVolume
        if (volumeUp) {
            earBackVolume += 1
        } else {
            earBackVolume -= 1
        }
        if (earBackVolume > 100) earBackVolume = 100
        if (earBackVolume < 0) earBackVolume = 0
        if (earBackVolume != mSetting.earBackVolume) {
            mSetting.earBackVolume = earBackVolume
        }
        binding.sbEarBack.progress = earBackVolume
    }

    private fun showSwitchEarModeDialog(mode: Int) {
        val cxt = context?:return
        if (switchEarModeDialog == null) {
            switchEarModeDialog = CommonDialog(cxt).apply {
                setDialogTitle("提示")
                if (mode == 1) {
                    setDescText("切换后将强制使用OpenSL模式，\n确认？")
                } else if (mode == 2) {
                    setDescText("切换后将强制使用Oboe模式，\n确认？")
                }
                setDialogBtnText(
                    getString(R.string.cantata_cancel),
                    getString(R.string.cantata_confirm)
                )
                onButtonClickListener = object : OnButtonClickListener {
                    override fun onLeftButtonClick() {
                        when (mSetting.earBackMode) {
                            0 -> {
                                this@EarBackFragment.binding.rgMode.check(R.id.tvModeAuto)
                            }
                            1 -> {
                                this@EarBackFragment.binding.rgMode.check(R.id.tvModeOpenSL)
                            }
                            else -> {
                                this@EarBackFragment.binding.rgMode.check(R.id.tvModeOboe)
                            }
                        }
                    }

                    override fun onRightButtonClick() {
                        mSetting.earBackMode = mode
                    }
                }
            }

        }
        switchEarModeDialog?.show()
    }


}